Remove block on editing switch expressions

davidwengier opened the pull request:
Fixes #43099

davidwengier commented on src/Compilers/CSharp/Test/Emit/PDB/PDBTests.cs:
```diff
@@ -4276,6 +4276,68 @@ .maxstack  2
+        <entry offset=""0x3b"" startLine=""11"" startColumn=""5"" endLine=""11"" endColumn=""6"" document=""1"" />
+      </sequencePoints>
+      <scope startOffset=""0x0"" endOffset=""0x3d"">
+        <scope startOffset=""0x1"" endOffset=""0x3b"">
+          <local name=""i"" il_index=""0"" il_start=""0x1"" il_end=""0x3b"" attributes=""0"" />
+          <scope startOffset=""0x16"" endOffset=""0x2b"">
+            <local name=""i"" il_index=""0"" il_start=""0x16"" il_end=""0x2b"" attributes=""0"" />
+          </scope>
```
The nested scopes with the same names here looks odd to me, but seems to make sense given (my interpretation of) the lowering code:

https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_SwitchExpression.cs#L102-L108

tmat replied:
It's indeed odd to have 2 IL local variables for "i". There is only one user-defined local "i".

---
In reply to: [552284955](https://github.com/dotnet/roslyn/pull/50270#discussion_r552284955) [](ancestors = 552284955)

tmat replied:
@jcouv Is this expected?

davidwengier replied:
https://github.com/dotnet/roslyn/issues/50301

tmat commented on src/Compilers/CSharp/Test/Emit/PDB/PDBLambdaTests.cs:
```diff
@@ -1296,6 +1296,156 @@ void F()
+    {
+        return o switch
+        {
+            int i => new Func<string>(() => $""Number: {i} + {o} == {i + (int)o}"")(),
```
>$""Number: {i} + {o} == {i + (int)o}"") [](start = 44, length = 39)

Nit: This could be a bit simpler (avoid string interpolations). It's better to simplify tests as much as possible to make it more focused on what's being tested and shorter. In this particular case we are not comparing IL, so it's ok.

tmat commented on src/Compilers/CSharp/Test/Emit/PDB/PDBLambdaTests.cs:
```diff
@@ -1296,6 +1296,156 @@ void F()
+}
+");
+
+            // we just want this to compile without crashing/asserting
```
>// we just want this to compile without crashing/asserting [](start = 12, length = 58)

I'd remove the comment. We do verify the generated PDB, so it's not just that it compiles.

tmat commented on src/EditorFeatures/CSharpTest/EditAndContinue/TopLevelEditingTests.cs:
```diff
@@ -3583,8 +3581,49 @@ class C
+            edits.VerifyEdits("Update [static int F(int a) => a switch { 0 => 0, _ => 1 };]@18 -> [static int F(int a) => a switch { 0 => 0, 1 => 1, _ => 2 };]@18");
+
+            edits.VerifyRudeDiagnostics();
         }
```
The tests in this file validate top-level matching. From top-level POV these additional tests do not test more than what MethodUpdate_UpdateSwitchExpression1 already does. It'd be better to move them to StatementEditingTests. I guess the first test was here to validate that Rude Edit is reported if there is any change in a method that contains switch expression, but I think that could also be moved to StatementEditingTests.

tmat commented on src/Features/Core/Portable/EditAndContinue/EditAndContinueDiagnosticDescriptors.cs:
```diff
@@ -135,8 +135,6 @@ void AddGeneralDiagnostic(EditAndContinueErrorCode code, string resourceName, Di
             AddRudeEdit(RudeEditKind.InsertLocalFunctionIntoInterfaceMethod, nameof(FeaturesResources.Adding_0_into_an_interface_method_will_prevent_the_debug_session_from_continuing));
             AddRudeEdit(RudeEditKind.InternalError, nameof(FeaturesResources.Modifying_source_file_will_prevent_the_debug_session_from_continuing_due_to_internal_error));
-            // TODO: remove (https://github.com/dotnet/roslyn/issues/43099)
-            AddRudeEdit(RudeEditKind.SwitchExpressionUpdate, nameof(FeaturesResources.Modifying_0_which_contains_a_switch_expression_will_prevent_the_debug_session_from_continuing));
```
>Modifying_0_which_contains_a_switch_expression_will_prevent_the_debug_session_from_continuing [](start = 86, length = 93)

We can clean up the resource as well.

tmat commented on src/Compilers/CSharp/Test/Emit/PDB/PDBLambdaTests.cs:
```diff
@@ -1296,6 +1296,156 @@ void F()
+        <encLambdaMap>
+          <methodOrdinal>0</methodOrdinal>
+          <closure offset=""0"" />
+          <closure offset=""20"" />
```
><closure offset=""20"" /> [](start = 10, length = 25)

Is offset 20 where the switch expression starts? It would be interesting to have nested switch expressions with closures. Will each switch expression have its own closure?

davidwengier replied:
Yes, offset 20 is the start of the switch. This was a good suggestion though, I think maybe it's unearthed another bug. Will comment below.

davidwengier commented on src/Compilers/CSharp/Test/Emit/PDB/PDBLambdaTests.cs:
```diff
@@ -1296,6 +1296,295 @@ void F()
+        </encLocalSlotMap>
+        <encLambdaMap>
+          <methodOrdinal>0</methodOrdinal>
+          <closure offset=""174"" />
```
This looks like another bug to me. Only one closure, at offset 174, which is the start of the nested switch. The lambda offsets looks right though, so maybe I'm interpreting it wrong though, because to me it seems like it would always be invalid to have a closure that starts after a lambda that creates it.

tmat replied:
Yeah, this looks wrong. Let's file another issue.

---
In reply to: [553091837](https://github.com/dotnet/roslyn/pull/50270#discussion_r553091837) [](ancestors = 553091837)

davidwengier replied:
https://github.com/dotnet/roslyn/issues/50321

jcouv commented on src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueClosureTests.cs:
```diff
@@ -125,6 +125,107 @@ void F()
                 Row(5, TableIndex.MethodDef, EditAndContinueOperation.Default));
         }
 
+        [Fact(Skip = "https://github.com/dotnet/roslyn/issues/43099")]
```
I'm working on a fix. Check whether this test can be unskipped when this PR is ready to merge.

davidwengier commented:
Rebased and updated one test now that https://github.com/dotnet/roslyn/issues/50301 is fixed. https://github.com/dotnet/roslyn/issues/50321 is still outstanding.

cston commented on src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueClosureTests.cs:
```diff
@@ -125,6 +125,107 @@ void F()
+                generation0,
+                ImmutableArray.Create(SemanticEdit.Create(SemanticEditKind.Update, f0, f1, GetSyntaxMapFromMarkers(source0, source1), preserveLocalVariables: true)));
+
+            // TODO: Below is incorrect and needs updating once compiler bugs are fixed
```
>TODO [](http://example.com/codeflow?start=15&length=4)

Is there an issue id that could be used here and below instead of TODOs?

davidwengier replied:
Oops, forgot about these. Thank you for noticing!

These need to be fixed in this PR, but didn't want to do so until the compiler bugs were fixed so we could be sure the closure info is correct. Putting this back to draft to work on it some more!

davidwengier commented:
@tmat any other feedback?
